### PR TITLE
[0.17] Add Elements-style witness structure

### PR DESCRIPTION
**Pull request by stevenroose**
Ports the Elements-style transaction witness structure from v0.14.1. Additionally, the Elements-compatible serialization of the witness data is gated behind a new `-con_elementswitness` flag.

Fixes https://github.com/ElementsProject/elements/issues/479.

### CODE REVIEW

File: src/primitives/transaction.h [hunk 1]
```diff
@@ -295,6 +297,9 @@ inline void UnserializeTransaction(TxType& tx, Stream& s) {
     unsigned char flags = 0;
     tx.vin.clear();
     tx.vout.clear();
+    tx.witness.SetNull();
+    const_cast<CTxWitness*>(&tx.witness)->vtxinwit.resize(tx.vin.size());
```
Comment (instagibbs) [inline src/primitives/transaction.h]: wrong place, will always resize to 0... move to right before `g_con_elementswitness` check

File: src/primitives/transaction.h [hunk 2]
```diff
@@ -295,6 +297,9 @@ inline void UnserializeTransaction(TxType& tx, Stream& s) {
...
+    const_cast<CTxWitness*>(&tx.witness)->vtxoutwit.resize(tx.vout.size());
```
Comment (instagibbs) [inline src/primitives/transaction.h]: wrong place, will always resize to 0... move to right before g_con_elementswitness check

**Comment by instagibbs** [comment]
needs rebase

File: test/functional/test_framework/messages.py [hunk 3]
```diff
@@ -474,8 +474,8 @@ def deserialize(self, f):
     # Only applicable for non-CT, non-segwit transactions
     def serialize_without_witness(self):
         r = b""
-        r += struct.pack("B", 1)
         r += struct.pack("<i", self.nVersion)
+        r += struct.pack("B", 1)
```
Comment (instagibbs) [inline test/functional/test_framework/messages.py]: reminder: this should probably be 0

File: test/functional/feature_txwitness.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        self.num_nodes = 1
+
+    def test_transaction_serialization(self):
+        node = self.nodes[0]
+
+        # spend money to segwit address
+        unspent = node.listunspent()[0]
+        addr = node.getnewaddress("", "p2sh-segwit")
+        node.sendtoaddress(address=addr, amount=50, subtractfeefromamount=True)
+        node.generate(1)
```
Comment (instagibbs) [inline test/functional/feature_txwitness.py]: let's use `generatetoaddress` instead, `generate` is removed upstream. You can send funds directly to any type of address you want

**Comment by stevenroose** [comment]
Added your fixes and also fixed benchmarks. Travis should pass now.

**Comment by instagibbs** [comment]
needs rebase ;)

**Comment by instagibbs** [comment]
```
src/consensus/validation.h:111: cound  ==> could, count
^ Warning: codespell identified likely spelling errors. Any false positives? Add them to the list of ignored words in test/lint/lint-spelling.ignore-words.txt
^---- failure generated from test/lint/lint-spelling.sh
```

Very important linting error that stopped builds

File: test/functional/feature_bip68_sequence.py [hunk 5]
```diff
@@ -371,7 +373,6 @@ def test_bip68_not_consensus(self):
         block.rehash()
         add_witness_commitment(block)
         block.solve()
-
```
Comment (instagibbs) [inline test/functional/feature_bip68_sequence.py]: random deletion

**Comment by instagibbs** [comment]
`DEBUG=1` build is timing out, just too slow to finish. Not a blocker.

**Comment by instagibbs** [comment]
getting a failure 50% of the time locally on `DEBUG=1` build for `./test/functional/wallet_txn_clone.py --segwit`, investigating

**Comment by instagibbs** [comment]
This fixes my test failure. Two things: 1) `40` vs `40*COIN` and 2) You weren't actually using the mutated transaction after flipping the outputs.

```
diff --git a/test/functional/wallet_txn_clone.py b/test/functional/wallet_txn_clone.py
index 0b90061..94c7f37 100755
--- a/test/functional/wallet_txn_clone.py
+++ b/test/functional/wallet_txn_clone.py
@@ -6,7 +6,7 @@
 
 import io
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.messages import CTransaction
+from test_framework.messages import CTransaction, COIN
 from test_framework.util import (
     assert_equal,
     connect_nodes,
@@ -77,12 +77,12 @@ class TxnMallTest(BitcoinTestFramework):
         # createrawtransaction randomizes the order of its outputs, so swap them if necessary.
         clone_tx = CTransaction()
         clone_tx.deserialize(io.BytesIO(hex_str_to_bytes(clone_raw)))
-        if (rawtx1["vout"][0]["value"] == 40 and clone_tx.vout[0].nValue != 40 or rawtx1["vout"][0]["value"] != 40 and clone_tx.vout[0].nValue == 40):
+        if (rawtx1["vout"][0]["value"] == 40 and clone_tx.vout[0].nValue != 40*COIN or rawtx1["vout"][0]["value"] != 40 and clone_tx.vout[0].nValue == 40*COIN):
             (clone_tx.vout[0], clone_tx.vout[1]) = (clone_tx.vout[1], clone_tx.vout[0])
 
         # Use a different signature hash type to sign.  This creates an equivalent but malleated clone.
         # Don't send the clone anywhere yet
-        tx1_clone = self.nodes[0].signrawtransactionwithwallet(clone_raw, None, "ALL|ANYONECANPAY")
+        tx1_clone = self.nodes[0].signrawtransactionwithwallet(clone_tx.serialize().hex(), None, "ALL|ANYONECANPAY")
         assert_equal(tx1_clone["complete"], True)
 
         # Have node0 mine a block, if requested:
```

**Comment by instagibbs** [comment]
ACK https://github.com/ElementsProject/elements/pull/494/commits/e3fe62b8e1b9ef2b10d8d44d5ee8d125c02121e8

Will merge after Travis finishes.

**Comment by instagibbs** [comment]
reACK https://github.com/ElementsProject/elements/pull/494/commits/52c8601cf4ef179cb522044dba5eddafc5a9e60d 

linter error fixes only change

**Comment by instagibbs** [comment]
Optimistic merging to allow CA rebase

**Pull request merged**